theano/gpuarray/dnn.py

```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        node = Apply(self, [dimensions],
+                     [CDataType("cudnnSpatialTransformerDescriptor_t",
+                      freefunc="cudnnDestroySpatialTransformerDescriptor")()])
+        # DebugMode cannot compare the values of CDataType variables, so by
+        # default it returns False all the time. To prevent DebugMode from
+        # complaining because of the MergeOptimizer, we make this variable
+        # always compare to True.
+        out = node.outputs[0]
+        out.tag.values_eq_approx = tensor.type.values_eq_approx_always_true
+        return node
+
+    def c_code_cache_version(self):
+        return (super(GpuDnnTransformerDesc, self).c_code_cache_version(), version())
+
+
+class GpuDnnTransformer(DnnBase):
+    """
+    Spatial transformer that can be used in spatial transformer networks, it
+    implements the grid generator and sampler. The localization network can
+    be built using neural net components of Theano.
+    """
+    __props__ = ()
+    _cop_num_inputs = 4
+    _cop_num_outputs = 2
+    _f16_ok = True
+    default_output = 0
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/dnn_sptf.c"], "APPLY_SPECIFIC(dnn_sptf)")
+
+    def make_node(self, img, theta, desc):
+        context_name = infer_context_name(desc)
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        img = gpu_contiguous(as_gpuarray_variable(img, context_name))
+        if img.type.ndim != 4:
+            raise TypeError('img must be a 4D tensor')
+        elif img.dtype not in ('float16', 'float32', 'float64'):
+            raise TypeError('img type must be floating-point')
+
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+        assert theta.dtype in ('float16', 'float32', 'float64')
+        assert theta.ndim == 3
+
+        # Setup grid dimensions using input from descriptor
+        grid_dims = as_tensor_variable(desc.owner.inputs[0])
+        assert grid_dims.dtype in theano.tensor.basic.int_dtypes
+        assert grid_dims.ndim == 1
+        # Ensure 64-bit ints are passed to the C code
+        grid_dims = theano.tensor.basic.cast(grid_dims, 'int64')
+
+        output = GpuArrayType(dtype=img.dtype,
+                              broadcastable=img.type.ndim * (False,),
+                              context_name=context_name)()
+
+        grid = GpuArrayType(dtype=img.dtype,
+                            broadcastable=img.type.ndim * (False,),
+                            context_name=context_name)()
+
+        inputs = [img, theta, grid_dims, desc]
+        outputs = [output, grid]
+        return Apply(self, inputs, outputs)
+
+    def L_op(self, inputs, outputs, grads):
+        img, theta, grid_dims, desc = inputs
+        _, grid = outputs
+        dy = grads[0]
```

Comment:
Hello @notoraptor , the issue I've mentioned, of converting an object to a ``GpuArrayVariable`` is caused here, because dy is a ``TensorVariable``. ``dy`` is passed to ``L_op`` through the ``verify_grad`` function. Is there a way to convert a ``TensorVariable`` to a ``GpuArrayVariable``?